### PR TITLE
Support for App Protect module

### PR DESCRIPTION
**Pull request by rafwegv**
### Proposed changes
Add App Protect integration to Ingress Controller. 
This adds two crd's and changes in go code that enables watching on those resources.
policies and log configuration can be added to ingresses using annotations.
CR's are read and written onto container storage as files , referenced in nginx configurations.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by Rulox** [comment]
Let's wait until #1009 is merged and make the changes here, I think is more fair/easier.

File: deployments/common/ap-policy-definition.yaml [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+                    type: object
+                  entityChanges:
+                    properties:
+                      type:
+                        type: string
+                    type: object
+                type: object
+              type: array
+            policy:
+              description: Foo is an example field of APPolicyConfig. Edit APPolicyConfig_types.go
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:56]: Fill in real description or remove?

File: deployments/common/ap-policy-definition.yaml [hunk 2]
```diff
@@ -0,0 +1,874 @@
...
+                            - Bad HTTP version
+                            - Bad host header value
+                            type: string
+                          enabled:
+                            type: boolean
+                          maxHeaders:
+                            description: optional
+                            type: integer
+                          maxParams:
+                            type: integer
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:146]: maxHeaders, maxParams, and evasions/maxDecoding passes are all "optional" as they only apply to one of the subviolation.

File: deployments/common/ap-policy-definition.yaml [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+                        format: int32
+                        maximum: 999
+                        minimum: 100
+                        type: integer
+                      type: array
+                    customXffHeaders:
+                      items:
+                        type: string
+                      type: array
+                    enforcementReadinessPeriod:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:405]: Not relevant for NAP as staging/readiness is not supported.

File: deployments/common/ap-policy-definition.yaml [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+                      checkSignatures:
+                        type: boolean
+                      htmlNormalization:
+                        type: boolean
+                      id:
+                        type: string
+                      kind:
+                        type: string
+                      lastUpdateMicros:
+                        type: integer
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:432]: # Fields that are from BIG-IP REST that are invalid or irrelevant for Declarative
- id
- kind
- lastUpdateMicros
- selfLink

File: deployments/common/ap-policy-definition.yaml [hunk 5]
```diff
@@ -0,0 +1,874 @@
...
+                        type: string
+                      isBase64:
+                        type: boolean
+                    type: object
+                  type: array
+                methods:
+                  items:
+                    properties:
+                      name:
+                        enum:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:496]: Custom methods can be added that are not defined in the RFC. They should not be limited to an enum list.

File: deployments/common/ap-policy-definition.yaml [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+                        type: boolean
+                      level:
+                        enum:
+                        - Global
+                        type: string
+                      metacharsOnParameterValueCheck:
+                        type: boolean
+                      name:
+                        enum:
+                        - '*'
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:531]: Parameter name can be values other than "*" now.

File: deployments/common/ap-policy-definition.yaml [hunk 7]
```diff
@@ -0,0 +1,874 @@
...
+                        type: boolean
+                      attackSignaturesCheck:
+                        type: boolean
+                      checkMaxValueLength:
+                        type: boolean
+                      checkMetachars:
+                        type: boolean
+                      level:
+                        enum:
+                        - Global
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:525]: "global" (must be lowercase)

File: deployments/common/ap-policy-definition.yaml [hunk 8]
```diff
@@ -0,0 +1,874 @@
...
+                        type: boolean
+                      name:
+                        type: string
+                      normalizationViolations:
+                        type: boolean
+                      percentDecoding:
+                        type: boolean
+                      selfLink:
+                        type: string
+                      type:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:445]: not a string. enum of wildcard|explicit

**Comment by Dean-Coakley** [comment]
> Let's wait until #1009 is merged and make the changes here, I think is more fair/easier.
> 
> EDIT: I'll follow up with Dean and coordinate who merges first, I don't want to delay this PR just because of the changes in Helm.

Just an update: I merged as agreed.

File: deployments/common/ap-policy-definition.yaml [hunk 9]
```diff
@@ -0,0 +1,874 @@
...
+                          tolerateJSONParsingWarnings:
+                            type: boolean
+                        type: object
+                      description:
+                        type: string
+                      hasValidationFiles:
+                        type: boolean
+                      name:
+                        enum:
+                        - Default
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:477]: Can have other names now.

File: deployments/common/ap-policy-definition.yaml [hunk 10]
```diff
@@ -0,0 +1,874 @@
...
+                          properties:
+                            isAllowed:
+                              type: boolean
+                            metachar:
+                              type: string
+                          type: object
+                        type: array
+                      parameterLocation:
+                        type: string
+                      performStaging:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:544]: Staging is not relevant for NAP

File: deployments/common/ap-policy-definition.yaml [hunk 11]
```diff
@@ -0,0 +1,874 @@
...
+                      signatureOverrides:
+                        items:
+                          properties:
+                            enabled:
+                              type: boolean
+                            signatureId:
+                              type: integer
+                          type: object
+                        type: array
+                      type:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:557]: not a string. enum of wildcard|explicit

File: deployments/common/ap-policy-definition.yaml [hunk 12]
```diff
@@ -0,0 +1,874 @@
...
+                  type: array
+                signature-sets:
+                  items:
+                    properties:
+                      alarm:
+                        type: boolean
+                      block:
+                        type: boolean
+                      name:
+                        enum:
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:715]: Custom signature sets are allowed to be defined, not just the enum list of factory sets.

File: deployments/common/ap-policy-definition.yaml [hunk 13]
```diff
@@ -0,0 +1,874 @@
...
+                  properties:
+                    name:
+                      type: string
+                  type: object
+                urls:
+                  items:
+                    properties:
+                      method:
+                        enum:
+                        - '*'
```
Comment (AvriBalofsky) [inline deployments/common/ap-policy-definition.yaml:768]: Custom urls are allowed to be defined, not just the pure wildcard

File: deployments/common/ap-policy-definition.yaml (hunk 8 above)
Comment (rafwegv) [reply deployments/common/ap-policy-definition.yaml:445]: which field are You refering to?

Comment (AvriBalofsky) [reply deployments/common/ap-policy-definition.yaml:445]: "type"
Just like for cookies and filetypes that you already have

**Comment by rafwegv** [comment]
I decided to add preserve unknown fields false to two fields, since they do not have a specific form (modifications) or are very complicated (signature sets).

**Comment by Rulox** [comment]
Closing in favour of https://github.com/nginxinc/kubernetes-ingress/pull/1035

**Pull request closed without merging**